benchmarks/renderImageItem.py

```diff
@@ -22,117 +23,156 @@ def renderQImage(*args, **kwargs):
-        self.uint8_data = None
-        self.uint8_lut = None
-        self.uint16_data = None
-        self.uint16_lut = None
-        self.cupy_uint16_lut = None
-        self.cupy_uint8_lut = None
-
-    def setup(self):
-        size = (self.size, self.size)
-        self.float_data, self.uint16_data, self.uint8_data, self.uint16_lut, self.uint8_lut = self._create_data(
-            size, np
-        )
-        if numba is not None:
+        self.data = np.empty((), dtype=np.uint8)
+        self.lut = np.empty((), dtype=np.ubyte)
+        self.levels = None
+
+    def teardown(self, *args, **kwargs):
+        # toggle options off
+        pg.setConfigOption("useNumba", False)
+        pg.setConfigOption("useCupy", False)
+
+    def setup_cache(self) -> dict:
+        accelerations = [np]
+        if cp is not None:
+            accelerations.append(cp)
+
+        cache = {}
+
+        for xp in accelerations:
+            cache[xp.__name__] = {"lut": {}, "data": {}}
+
+            # handle lut caching
+            c_map = xp.array([[-500.0, 255.0], [-255.0, 255.0], [0.0, 500.0]])
+            for lut_dtype in self.params.lut_dtypes:
+                if lut_dtype is None:
+                    continue
+                bits = xp.dtype(lut_dtype).itemsize * 8
+                # create the LUT
+                lut = xp.zeros((2 ** bits, 4), dtype="ubyte")
+                for i in range(3):
+                    lut[:, i] = xp.clip(xp.linspace(c_map[i][0], c_map[i][1], 2 ** bits), 0, 255)
+                lut[:, -1] = 255
+                cache[xp.__name__]["lut"][lut_dtype] = lut
+
+            # handle data caching
+            for dtype in self.params.dtypes:
+                cache[xp.__name__]["data"][dtype] = {}
+                for size in self.params.sizes:
+                    if xp.dtype(dtype) in (xp.float32, xp.float64):
+                        data = xp.random.default_rng().standard_normal(size=size, dtype=dtype)
+                    else:
+                        iinfo = xp.iinfo(dtype)
+                        data = xp.random.default_rng().integers(
+                            low=iinfo.min,
+                            high=iinfo.max,
+                            size=size,
+                            dtype=dtype
+                        )
+                    cache[xp.__name__]["data"][dtype][size] = data
+        return cache
+
+
+    def setup(
+            self,
+            cache: dict,
+            size: tuple[int, int],
+            acceleration: str,
+            use_levels: bool,
+            dtype: npt.DTypeLike,
+            lut_dtype: typing.Optional[npt.DTypeLike]
+    ):
+        xp = np
+        self.name = f"{size=}_{acceleration=}_{use_levels=}_{dtype=}_{lut_dtype=}"
+        if acceleration == "numba":
+            if numba is None:
+                raise NotImplementedError
+            if lut_dtype is None and not use_levels:
+                # this is a fast code-path bypassing makeARGB
```

Comment:
This comment is not correct. The parameters in this benchmark script should (hopefully) never invoke `makeARGB`. If the format of the rendered `QImage` is `ARGB32`, then it has invoked `makeARGB`.

Reply:
Thanks for pointing this out; I decided to skip the tests that are able to be rendered to QImage directly if running on non-numpy acceleration permutations of the parameters; as there is no point in testing that.